### PR TITLE
[API-377] Add reliable topic cluster restart tests

### LINKED ISSUE #758: [TRACKING ISSUE] Fix ClientReliableTopicOnClusterRestartTest
The tracking issue for the Java side PR.

**Fix ClientReliableTopicOnClusterRestartTest**

See https://github.com/hazelcast/hazelcast/pull/16644 for details.

---

Fixes the test testing behaviour of a loss tolerant listener when the
instance is restarted. This test failure is similar to the one described
in https://github.com/hazelcast/hazelcast/issues/16430#issuecomment-571947692 and the [fix PR](https://github.com/hazelcast/hazelcast/pull/16440).

Basically, it might happen that the new item is added to the restarted
member before the listener sends a read operation to the new member.
Since it was previously on a higher sequence, the sequence will be
reset to the currently highest, which is one after the latest added item.

In the case when the item is added to the ringbuffer before the read
operation is executed, the listener will simply ignore the added item
and wait for a new one.

The test asserts that the listener is able to continue after the cluster
restarts so we modify the test so that we add new items until we
eventually get an item.

Also, made some minor modifications like using small instances in the
test and made the sequence field volatile since it is accessed from
different threads.

Fixes: https://github.com/hazelcast/hazelcast/issues/16623

### PR DESCRIPTION
**Pull request by srknzl**
fixes #758 

I added `shouldContinue_OnClusterRestart_whenDataLoss_LossTolerant_afterInvocationTimeout`, should we add more tests here?

### CODE REVIEW

**Comment by srknzl** [comment]
verify

File: test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    });
+
+    after(async function() {
+        await RC.shutdownCluster(cluster.id);
+    });
+
+    it('should continue on cluster restart when data lost and after invocation timeout', async function () {
+        const invocationTimeoutMillis = 2000;
+
+        client1 = await Client.newHazelcastClient({
```
Comment (mdumandag) [inline test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js:54]: I think we can move client construction to a function to eliminate duplicate code

File: test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        // wait some time for subscription
+        await promiseWaitMilliseconds(invocationTimeoutMillis);
+
+        await assertTrueEventually(async () => {
+            const topic3 = await client2.getReliableTopic(topicName);
+            await topic3.publish(`newItem${randomString()}`);
+            await assertTrueEventually(async () => {
+                if (!messageArrived)
+                    throw new Error('message not arrived yet');
+            }, undefined, 5000);
```
Comment (mdumandag) [inline test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js:104]:
```suggestion
            await assertTrueEventually(() => messageArrived.should.be.true, undefined, 5000);
```

File: test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    afterEach(async function () {
+        if (client1) await client1.shutdown();
+        if (client2) await client2.shutdown();
+    });
+
+    after(async function() {
+        await RC.shutdownCluster(cluster.id);
+    });
+
+    it('should continue on cluster restart when data lost and after invocation timeout', async function () {
```
Comment (mdumandag) [inline test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js:51]:
```suggestion
    it('should continue on cluster restart when data loss and after invocation timeout', async function () {
```

File: test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            connectionStrategy: {
+                connectionRetry: {
+                    clusterConnectTimeoutMillis: Number.MAX_SAFE_INTEGER
+                }
+            }
+        });
+
+        const topic2 = await client2.getReliableTopic(topicName);
+        await topic2.publish('message');
+        await assertTrueEventually(async () => {
```
Comment (mdumandag) [inline test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js:151]: similar thing as above

File: test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+        topic1.addMessageListener(() => {
+            messageArrived = true;
+        });
+
+        await RC.shutdownMember(cluster.id, member.uuid);
+        await RC.startMember(cluster.id);
+
+        await topic2.publish('message');
+        await assertTrueEventually(async () => {
```
Comment (mdumandag) [inline test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js:189]: similar thing as above

Comment (srknzl) [reply test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js]: these are done in https://github.com/hazelcast/hazelcast-nodejs-client/pull/890/commits/c6f8d1483c1ace172ab81f96dab1e495382cc8ad

**Review by mdumandag — APPROVED** [review]

File: test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        member = await RC.startMember(cluster.id);
+    });
+
+    beforeEach(async function () {
+        client1 = undefined;
+        client2 = undefined;
+    });
+
+    afterEach(async function () {
+        if (client1) await client1.shutdown();
```
Comment (mdumandag) [inline test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js:43]: BTW, can you enable https://eslint.org/docs/rules/curly and https://eslint.org/docs/rules/keyword-spacing? So, we don't have such statements

Comment (srknzl) [reply test/integration/backward_compatible/topic/ReliableTopicOnClusterRestart.js:43]: ok will do in a seperate pr, opening an issue

**Pull request merged**